pkg/webhook/util/writer/external.go

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright 2020 The Kruise Authors.
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package writer
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	clientset "k8s.io/client-go/kubernetes"
+
+	"github.com/openkruise/kruise/pkg/webhook/util/generator"
+)
+
+const (
+	ExternalCertWriter = "external"
+	ExternalCACert     = "ca.crt"
+	ExternalCAKey      = "ca.key"
+	ExternalServerCert = "tls.crt"
+	ExternalServerKey  = "tls.key"
+)
+
+// externalCertWriter provisions the certificate by reading from the k8s secrets.
+type externalCertWriter struct {
+	*ExternalCertWriterOptions
+
+	// dnsName is the DNS name that the certificate is for.
+	dnsName string
+}
+
+// ExternalCertWriterOptions is options for constructing a externalCertWriter.
+type ExternalCertWriterOptions struct {
+	// client talks to a kubernetes cluster for creating the secret.
+	Clientset clientset.Interface
+	// secret points the secret that contains certificates that written by the CertWriter.
+	Secret *types.NamespacedName
+}
+
+var _ CertWriter = &externalCertWriter{}
+
+func (ops *ExternalCertWriterOptions) validate() error {
+	if ops.Clientset == nil {
+		return errors.New("client must be set in externalCertWriterOptions")
+	}
+	if ops.Secret == nil {
+		return errors.New("secret must be set in externalCertWriterOptions")
+	}
+	return nil
+}
+
+// NewExternalCertWriter constructs a CertWriter that persists the certificate in a k8s secret.
+func NewExternalCertWriter(ops ExternalCertWriterOptions) (CertWriter, error) {
+	err := ops.validate()
+	if err != nil {
+		return nil, err
+	}
+	return &externalCertWriter{ExternalCertWriterOptions: &ops}, nil
+}
+
+// EnsureCert read and validate certs from k8s secret.
+func (s *externalCertWriter) EnsureCert(dnsName string) (*generator.Artifacts, bool, error) {
+	// Create or refresh the certs based on clientConfig
+	s.dnsName = dnsName
+	certs, err := s.read()
+	if err != nil {
```

Comment:
is it possible the ca read from the secret is the same as before ? it necessary to update webhookconfiguration again ?

Reply:
add logic to check if the certs are same as before, however, the `changed` bool value is not used in webhook controller, and the webhook controller will use `reflect.DeepEqual()` to see if the webhook configuration need to be updated.